active/0000-engines.md

```diff
@@ -0,0 +1,411 @@
+ownership and encapsulation. Both will also have initializers and instance
+initializers.
+
+Engine instances will have access to their parent instances. An engine's parent
+could be either an application or engine.
+
+#### Routable vs. route-less engines
+
+Routable engines will define their routes in a new `Ember.Routes` class. This
+class will encapsulate the functionality provided by `Router#map`, and will be
+used internally by `Ember.Router` as well (with no public interface changes of
+course).
+
+Route-less engines do not define routing maps nor can they contain routes.
+
+### Developing engines
+
+Engines can be developed in isolation as Ember CLI addon projects or as part of
+a parent application.
+
+#### Engines as addons
+
+Engines can be created as separate addon projects with:
+
+```
+ember engine <engine-name>
+```
+
+This will create a special form of an ember addon. The file structure will match
+that of a standard addon, but will have an `engine` directory instead of an
+`addon` directory.
+
+Engines can be unit tested and can also be integration tested within a dummy
+app, just like standard addons.
+
+#### In-repo engines
+
+An engine can be created within an existing application's project using a
+special `in-repo-engine` generator (similar to the `in-repo-addon` generator):
+
+```
+ember g in-repo-engine <engine-name>
+```
+
+In-repo engines can be unit tested in isolation or integration testing with the
+main application (instead of a dummy application).
+
+> Note: In-repo addons currently are created in the `/lib` directory (e.g.
+`/lib/my-addon`). Unit tests and integration tests are currently co-mingled with
+tests for the main application. It's recommended that in-repo engines provide
+better test separation than is provided for regular addons, and perhaps the
+whole in-repo addon directory structure should be re-examined at the same time
+in-repo engines are introduced.
+
+#### Engine directory structure
+
+An engine's directory will contain a file structure identical to the `app`
+directory in a standard ember-cli application, with the following exceptions:
+
+* `engine.js` instead of `app.js` - defines the `Engine` class and
+  loads its initializers.
+
+* `routes.js` instead of `router.js` - defines an engine's routing map in a
+  `Routes` class. This file should be deleted entirely for route-less engines.
+
+### Installing engines
+
+Engines developed as addons can be installed in an application just like any
+other addon:
+
+```
+ember install <engine-name>
+```
+
+During development, you can use `npm link` to make your engine available in
+another parent engine or application.
+
+### Mounting routable engines
+
+The new `mount()` router DSL method is used to mount an engine at a particular
+"mount-point" in a route map.
+
+For example, the following route map mounts the `discourse` engine at the
+`/forum` path:
+
+```
+Router.map(function() {
+  this.mount('discourse', {path: '/forum'});
```

Comment:
I'm assuming we can `mount` anywhere?

Can we mount within another mount in this DSL, or are nested engines left to the internal route DSL of a given engine.